test/jdk/com/sun/jndi/ldap/LdapSSLHandshakeFailureTest.java

```diff
@@ -0,0 +1,205 @@
+
+    public static void main(String args[]) throws Exception {
+        // Set the keystores
+        setKeyStore();
+        // start the test server first.
+        TestServer server = new TestServer();
+        server.start();
+        Hashtable<String, Object> env = new Hashtable<>();
+        env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");
+        env.put(Context.PROVIDER_URL, URIBuilder.newBuilder()
+                .scheme("ldaps")
+                .loopback()
+                .port(server.getPortNumber())
+                .buildUnchecked().toString());
+        env.put("java.naming.ldap.factory.socket", CustomSocketFactory.class.getName());
+        env.put("java.naming.ldap.version", "3");
+        env.put("com.sun.jndi.ldap.connect.timeout", "1000");
+        env.put(Context.SECURITY_AUTHENTICATION, "Simple");
+        env.put(Context.SECURITY_PRINCIPAL, "cn=principal");
+        env.put(Context.SECURITY_CREDENTIALS, "justpassword");
+        LdapContext ctx = null;
+        try {
+            ctx  = new InitialLdapContext(env, null);
+        } catch (Exception e) {
+            if (CustomSocketFactory.customSocket.closeMethodCalledCount() > 0) {
+                System.out.println(SOCKET_CLOSED_MSG);
+            } else {
+                System.out.println(SOCKET_NOT_CLOSED_MSG);
+                throw e;
+            }
+        } finally {
+            if(ctx != null)
+                ctx.close();
+        }
+    }
+    public static class CustomSocketFactory extends SocketFactory {
+        public static CustomSocket customSocket = new CustomSocket();
+
+        public static CustomSocketFactory getDefault() {
+            return new CustomSocketFactory();
+        }
+
+        @Override
+        public Socket createSocket() {
+            return customSocket;
+        }
+
+        @Override
+        public Socket createSocket(String s, int timeout) {
+            return customSocket;
+        }
+
+        @Override
+        public Socket createSocket(String host, int port, InetAddress localHost,
+                                   int localPort) {
+            return customSocket;
+        }
+
+        @Override
+        public Socket createSocket(InetAddress host, int port) {
+            return customSocket;
+        }
+
+        @Override
+        public Socket createSocket(InetAddress address, int port,
+                                   InetAddress localAddress, int localPort) {
+            return customSocket;
+        }
+    }
+
+    private static class CustomSocket extends Socket {
+        private int closeMethodCalled = 0;
+
+        public int closeMethodCalledCount() {
+            return closeMethodCalled;
+        }
+
+        @Override
+        public void close() throws IOException {
+            closeMethodCalled++;
+            super.close();
+        }
+    }
+
+    private static void setKeyStore() {
```

Comment:
Overall code changes looks OK to me,  did you consider creating the self signed certificates programmatically and use them in  your test ?. Please  look into the existing javax.net.ssl tests how to do this.